File: leet/0015-3sum/3sum.go
import (
    "sort"
)

func threeSum(nums []int) [][]int {
    result := make([][]int, 0)
    
    if nums == nil || len(nums) < 3 {return result}
    
    sort.Ints(nums)
    length := len(nums)
    for i := 0; i < length; i++ {
        for j, k:= i + 1, length - 1; j < k; {
            sum := nums[j]+ nums[k] + nums[i]
            if sum == 0 {
                r1 := make([]int, 3)
                r1[0] = nums[i]
                r1[1] = nums[j]
                r1[2] = nums[k]
                result = append(result, r1)
                
                //j++
                for j + 1 < k && nums[j] == nums[j + 1] {j++}
                j++
                //k--
                for j < k - 1 && nums[k] == nums[k - 1] {k--}
                k--
            } else if (sum > 0){
                // k--
                for j < k - 1 && nums[k] == nums[k - 1] {k--}
                k--
            } else{
                // j++
                for j + 1 < k && nums[j] == nums[j + 1] {j++}
                j++
            }
        }
        
        for i + 1 < length && nums[i] == nums[i + 1] {i++}
    }
    
    return result
}

File: leet/0015-3sum/3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    if (nums.length < 3) return []

    let ret = []
    nums.sort((a, b) => a - b)
    for (let i = 0; i < nums.length; i++){
        // loop
        let j = i + 1
        let k = nums.length - 1

        while(j < k){
            let sum = nums[i] + nums[j] + nums[k]
            if(sum == 0){
                ret.push([nums[i] , nums[j] , nums[k]])
                // ++ j
                while(j + 1 < k && nums[j] == nums[j + 1]) {j++}
                j++
                // -- k
                while(j < k - 1 && nums[k] == nums[k - 1]) {k--}
                k--
            } else if(sum < 0){
                // ++ j
                while(j + 1 < k && nums[j] == nums[j + 1]) {j++}
                j++
            } else {
                // -- k
                while(j < k - 1 && nums[k] == nums[k - 1]) {k--}
                k--
            }
        }

        while(i + 1< nums.length && nums[i] == nums[i + 1]) i += 1
    }

    return ret
};
